src/keyschedule.js

```diff
@@ -0,0 +1,96 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+'use strict';
+
+// TLS1.3 Key Schedule.
+//
+// In this file we implement the "key schedule" from
+// https://tools.ietf.org/html/rfc8446#section-7.1, which
+// defines how to calculate various keys as the handshake
+// state progresses.
+//
+
+
+import {
+  assert,
+  BufferWriter,
+  EMPTY,
+  zeros,
+} from './utils.js';
+import {
+  hkdfExtract,
+  hkdfExpandLabel,
+  HASH_LENGTH,
+  hash,
+  hmac,
+} from './crypto.js';
+
+const STAGE_UNINITIALIZED = 0;
+const STAGE_EARLY_SECRET = 1;
+const STAGE_HANDSHAKE_SECRET = 2;
+const STAGE_MASTER_SECRET = 3;
+
+export class KeySchedule {
+  constructor() {
+    // WebCrypto doesn't support a rolling hash construct, so we have to
+    // keep the entire message transcript in memory.
+    this.transcript = new BufferWriter();
+    // This tracks the main secret from with other keys are derived at each stage.
+    this.secret = null;
+    this.stage = STAGE_UNINITIALIZED;
+    // And these are all the various keys we'll derive as the handshake progresses.
+    this.extBinderKey = null;
+    this.clientHandshakeTrafficSecret = null;
+    this.serverHandshakeTrafficSecret = null;
+    this.clientApplicationTrafficSecret = null;
+    this.serverApplicationTrafficSecret = null;
+  }
+
+  async addPSK(psk) {
+    // Use the selected PSK (if any) to calculate the "early secret".
+    psk = psk || zeros(HASH_LENGTH);
+    assert(this.stage === STAGE_UNINITIALIZED, 'PSK added at incorrect state');
+    this.secret = await hkdfExtract(zeros(HASH_LENGTH), psk);
+    this.stage = STAGE_EARLY_SECRET;
+    this.extBinderKey = await this.deriveSecret('ext binder', EMPTY);
+  }
+
+  async addECDHE(ecdhe) {
```

Comment:
can this be `ecdhe  = zeros(HASH_LENGTH)`?

Reply:
I would prefer to make the caller pass in an explicit "I don't have an ECDHE output" indicator, seems safer to me. I think I'll make this a more explicit check for `null` rather than just falsiness.